Extract Groonga::Log::Parser from fluent-plugin-groonga-log

kenhys opened the pull request:
This PR is aimed to split fluent-plugin-groonga-log architecture into two parts.

1. fluent-plugin-groonga-log (use groonga-log parser library)
2. groonga-log (provide parser as a library)

kenhys committed:
Enable "rake test" target

kenhys committed:
Add missing dependency to test-unit gems

kou commented on Rakefile:
```diff
@@ -1,2 +1,14 @@
 require "bundler/gem_tasks"
 task :default => :spec
```
It should be changed to `:test`.

kou commented on lib/groonga/log/parser.rb:
```diff
@@ -0,0 +1,63 @@
+
+module Groonga
+  module Log
+    class Parser
```
`GroongaLog::Parser` is preferred to namespace conflict with Rroonga.

kou commented on lib/groonga/log/parser.rb:
```diff
@@ -0,0 +1,63 @@
+module Groonga
+  module Log
+    class Parser
+      REGEXP =
```
Need more meaningful name.

kou commented on lib/groonga/log/parser.rb:
```diff
@@ -0,0 +1,63 @@
+          next unless line.valid_encoding?
+          m = REGEXP.match(line)
+          unless m
+            yield nil, nil
```
Why?

kou commented on lib/groonga/log/parser.rb:
```diff
@@ -0,0 +1,63 @@
+            "hour" => hour,
+            "minutes" => minutes,
+            "seconds" => seconds,
+            "micro_seconds" => micro_seconds,
```
They should be number instead of string.

kou commented on lib/groonga/log/parser.rb:
```diff
@@ -0,0 +1,63 @@
+            "minutes" => minutes,
+            "seconds" => seconds,
+            "micro_seconds" => micro_seconds,
+            "log_level" => log_level,
```
How about normalize log level such as `:notice` instead of `"n"`?

kou commented on test/helper.rb:
```diff
@@ -0,0 +1,2 @@
+require "stringio"
```
License header is missing.

kou commented on test/test-parser.rb:
```diff
@@ -0,0 +1,38 @@
+require "groonga/log/parser"
+
+class ParserTest < Test::Unit::TestCase
+
```
A needless empty line.

kou commented on test/test-parser.rb:
```diff
@@ -0,0 +1,38 @@
+require "helper"
```
License header is missing.

kou commented on test/test-parser.rb:
```diff
@@ -0,0 +1,38 @@
+end
+
+
+
```
Needless empty lines.

kou commented on lib/groonga/log/parser.rb:
```diff
@@ -0,0 +1,63 @@
+        /\A(?<year>\d{4})-(?<month>\d\d)-(?<day>\d\d) (?<hour>\d\d):(?<minutes>\d\d):(?<seconds>\d\d)\.(?<micro_seconds>\d+)\|(?<log_level>.)\|(?<context_id>.+?)\|(?<message>.*)/
+
+      def parse(input, &block)
+        input.each_line do |line|
```
How about adding `return to_enum(:parse, input) unless block_given?)`?
We can write `parser.parse(input).to_a` with it.

kou commented on lib/groonga/log/parser.rb:
```diff
@@ -0,0 +1,63 @@
+      REGEXP =
+        /\A(?<year>\d{4})-(?<month>\d\d)-(?<day>\d\d) (?<hour>\d\d):(?<minutes>\d\d):(?<seconds>\d\d)\.(?<micro_seconds>\d+)\|(?<log_level>.)\|(?<context_id>.+?)\|(?<message>.*)/
+
+      def parse(input, &block)
```
`&block` isn't used.

kenhys committed:
Set default task to :test

kenhys commented:
* set :test as default rake task
* fix to `Groonga::GroongaLog::Parser` module naming
* use more meaningful regex pattern name
* apply `to_enum` if block is not given.
* return number instead of string expression for each timestamp
* add missing license header
* remove needless empty lines.

kou commented on lib/groonga/groonga-log.rb:
```diff
@@ -0,0 +1,8 @@
+require "groonga/groonga-log/version"
+require "groonga/groonga-log/parser"
+
+module Groonga
```
We don't need `Groonga` namespace.

kenhys replied:
I've followed groonga-query-log gem. It uses `Groonga` namespace.

kou replied:
groonga-query-log should be changed too.

kou commented on lib/groonga/groonga-log/parser.rb:
```diff
@@ -0,0 +1,86 @@
+module Groonga
+  module GroongaLog
+    class Parser
+      EXTRACT_GROONGA_LOG_REGEXP =
```
`FORMAT` or `PATTERN` is better.
`GROONGA_LOG` is redundant because we are in `GroongaLog` namespace.

kou commented on lib/groonga/groonga-log/parser.rb:
```diff
@@ -0,0 +1,86 @@
+          hour = m['hour'].to_i
+          minutes = m['minutes'].to_i
+          seconds = m['seconds'].to_i
+          micro_seconds = m['micro_seconds'].to_i
```
`Integer(xxx)` is better than `xxx.to_i`.

kou commented on lib/groonga/groonga-log/parser.rb:
```diff
@@ -0,0 +1,86 @@
+      def log_level_to_symbol(level_text)
+        case level_text
+        when "E"
+          return :emergency
```
`return` is redundant.

kou commented on test/test-parser.rb:
```diff
@@ -0,0 +1,89 @@
+  def parse(log)
+    statistics = []
+    parser = Groonga::GroongaLog::Parser.new
+    parser.parse(log) do |statistic|
```
`parser.parse(log).to_a`.

kou commented on test/test-parser.rb:
```diff
@@ -0,0 +1,89 @@
+    assert_equal(expected, statistics)
+  end
+
+  def test_to_a
```
It's needless because we can use `to_a` in `parse`.

kou commented on test/test-parser.rb:
```diff
@@ -0,0 +1,89 @@
+require "helper"
+
+class ParserTest < Test::Unit::TestCase
+
```
A needless empty line.

kenhys commented:
* Renamed regex name to  `PATTERN`
* Use `Interger(xxx)`
* Removed redundant `return`
* Removed needless `to_a` testcase. Use `parser().to_a` instead.
* Removed needless empty line.

cosmo0920 commented on lib/groonga/groonga-log/parser.rb:
```diff
@@ -0,0 +1,86 @@
+          log_level = log_level_to_symbol(m['log_level'])
+          context_id = m['context_id']
+          message = m['message']
+          timestamp = Time.local(year, month, day,
```
`Time.local` is always forcing users to use their local time zone.
Should we provide UTC or user specific time zone parsing option?
If Groonga log contents is always written with local timezone, this comment seems to be useless....

kou replied:
Groonga always uses local time.

cosmo0920 replied:
Hm, I see.... :thinking:

Fluentd plugins sometimes can choose UTC or local time for using time.
Current implementation looks good to me.

kenhys committed:
Put module under GroongaLog namespace

kenhys committed:
Implement GroongaLog::Parser

This implementation is extracted from fluent-plugin-groonga-log.

kenhys committed:
Add test case for GroongaLog::Parser

kenhys commented:
I've fixed namespace and rebased.

* Put module under GroongaLog namespace.

kenhys committed:
Fix gemspec namespace

kenhys commented:
@kou could you review again?

kou commented on lib/groonga-log.rb:
```diff
@@ -0,0 +1,6 @@
+
+module GroongaLog
+  # Your code goes here...
+end
```
This `module` is needless.

kou commented on test/run-test.rb:
```diff
@@ -0,0 +1,46 @@
+
+require "test-unit"
+require "test/unit/notify"
+require "test/unit/rr"
```
test-unit-rr is needless.

kou commented on lib/groonga-log/parser.rb:
```diff
@@ -0,0 +1,84 @@
+          "log_level" => log_level,
+          "context_id" => context_id,
+          "message" => message,
+        }
```
It's better that we use symbol instead of string for key such as `:timestamp`.
